### DISCUSSION TITLE
Dispose intermediate results of operators

### DISCUSSION THREAD

**Discussion by lellid** [dotnet/csharplang#3538] [Language Ideas]
* [x] Proposed
* [ ] Prototype: [Complete](https://github.com/PROTOTYPE_OWNER/roslyn/BRANCH_NAME)
* [ ] Implementation: [In Progress](https://github.com/dotnet/roslyn/BRANCH_NAME)
* [ ] Specification: [Not Started](pr/1)

## Summary

Intermediate results of arithmetic operations should be disposed, 
if the intermediate result implements the IDisposable interface, and a
special attribute will flag the result.

## Motivation
In mathematics with vectors and matrices, we often deal with large buffers,
which preferably come from a pool and must be returned to the pool if no longer in use.
In order to return the buffer to the pool, the Dispose method could be used. The problem
is that not in all situations there is a chance to call the Dispose method.

Consider the following code (class M stands for the matrix class):

```
 public class M : IDisposable
    {
        private double _x;

        public M(double x) { _x = x; Console.WriteLine($"Created: {_x}"); }

        public void Dispose() => Console.WriteLine($"Disposed: {_x}");

        public static M operator +(M a, M b)
        {
            return new M(a._x + b._x);
        }
    }


class Program
    {
    static void Main(string[] args)
        {
            using var a = new M(2);
            using var b = new M(3);
            using var c = new M(7);
            using var r = (a + b) + c;
        }
    }
```

This would result in the following output:

```
Created: 2
Created: 3
Created: 7
Created: 5
Created: 12
Disposed: 12
Disposed: 7
Disposed: 3
Disposed: 2
```

It is obvious that the intermediate result `(a+b)`, resulting in `5`, 
is not being disposed. Moreover, there is not even a chance to dispose it,
because it is not directly available in the source code.

## Detailed design

Add an attribute, for instance `DisposeIntermediateResultAttribute`. If applied to the return  value of an operator, the compiler should call `Dispose` after an intermediate result is no longer in scope.
This could look like this:

```
 public class M : IDisposable
    {
        private double _x;

        public M(double x) { _x = x; Console.WriteLine($"Created: {_x}"); }

        public void Dispose() => Console.WriteLine($"Disposed: {_x}");

        [return: DisposeIntermediateResult]
        public static M operator +(M a, M b)
        {
            return new M(a._x + b._x);
        }
    }
```

It is essential, that the Dispose method on the intermediate result(s) is directly called
after the intermediates are out of scope (at latest at the end of the source code line, and not just at the end of the method). 
This is because those expressions are often used in loops, and the buffer(s) are needed
again the next time those expression is evaluated. 

Since this is a new attribute, existing code should not be affected.

## Drawbacks

It is not clear yet what should happen if the Dispose method throws an exception.

## Alternatives

Alternatively, one could take the mathematical steps separately, for example (following the code from above):

```
static void Main(string[] args)
        {
            using var a = new M(2);
            using var b = new M(3);
            using var c = new M(7);
            using var intermediate = a + b;
            using var r = intermediate + c;
        }
    
```

This will result in ugly and unreadable code, especially if the expressions are more complex.

**Comment by HaloFour** [comment]
It seems quite contrary to the design of `using` which expects that the developer explicitly is opting-in to the disposable behavior and impact on lifecycle.  This silently injects code into the middle of expressions that the developer might never be aware of.  I know that's your intention, but that's why I don't like it anymore than I would like C# silently auto-disposing of any disposables.

**Comment by iam3yal** [comment]
@lellid 

You have to do the work yourself inside `M` by tracking `a` and `b` (or using a `Stack` but for the scenario above tracking `a` and `b` should suffice) depends on the complexity and then call `Dispose` on them inside `Dispose` so when the expression is `Dispose`d the instance that holds the sum will get disposed too.

> It is not clear yet what should happen if the Dispose method throws an exception.

_To help ensure that resources are always cleaned up appropriately, a Dispose method should be idempotent, such that it is callable multiple times without throwing an exception. Furthermore, subsequent invocations of Dispose should do nothing._ [Source](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose).

**Comment by lellid** [comment]
@eyalsk:

Yes, of course it is possible to do it by code. The compiler has to do it with code, too. But, hey, this is a language design forum. Language is supposed to make life easier!
In this simple case tracking as you described it might work, but in practice, a lot of other classes are involved, too (vectors, scalars, different kind of matrices). That may be the reason why I don't know of any C# matrix library that is using tracking.

@HaloFour:

I agree with you that disposing in the middle of an expression is not a good idea. And I agree with you also that silently disposing something is not a good design. Maybe (I didn't dare to propose that) one has to use the attribute in combination with a keyword, in a similar fashion for instance to the `await` keyword. Lets say such a keyword would be named `interdisp` (I am sure there is a better name for it).

Then the code could look like this:

```
class Program
    {
    static void Main(string[] args)
        {
            using var a = new M(2);
            using var b = new M(3);
            using var c = new M(7);
            using var r = interdisp (a + b) + c;
        }
    }
```

I'm fully aware that implementing this is not easy. For instance, how could the compiler know,
if calling the operator `+` with the intermediate result, that that function does not intent to store the intermediate result, and is only using it temporarily? That's why, the keyword `interdisp` might say: hey, I'm absolutely sure that all my operators and functions in the expression that follows will only use the arguments temporarily, and not store any of the intermediates permanently. All intermediates then could be disposed when the result is assigned to `r`, which then is a defined point.

**Comment by CyrusNajmabadi** [comment]
I would view it as an abuse of operator overloading to have them return disposable values that need disposing.  It would entirely break my expectations that this would be necessary and that i couldn't just treat temporary intermediary values as precisely that.  If these are complex objects needing cleanup, i would expect to use normal method calls, with very clearly defined lifetimes for the operations and the values produced from them.

**Comment by lellid** [comment]
Yes, and that may be a reason why C# is not so great in data science. For a scientist, there is a difference if she or he has to write `(a + b + c) * (e + f)`, or
```
var ab = Add(a, b);
var abc = Add(ab, c);
var ef = Add(e, f);
var result = Multiply(abc, ef);
```

You see the difference? If you would only have the latter source, could you easily see the expression behind it?

And disposing of the matrices is not really needed. The garbage collector will finally do the job. That's the current status of the available C# matrix libraries. Only, that is a little late, and the storage space could be used more efficiently by using Dispose to return the allocated storage space back to a pool. Maybe you could imagine what would happen if you use the above expression with 1000 x 1000 matrices in a loop with 100 iterations.

**Comment by CyrusNajmabadi** [comment]
>  If you would only have the latter source, could you easily see the expression behind it?

Yes :)

> Only, that is a little late, and the storage space could be used more efficiently by using Dispose to return the allocated storage space back to a pool. 

My approach would be to do things like this:

```c#
using var arena = CreateMathArena();
var a = arena.CreateM(2);
var b = arena.CreateM(3);
var c = arena.CreateM(7);
var r = (a + b) + c;
```

I would never expect or want the disposing to happen on each value, but across teh lifetime of the values that are all aggregated together.

**Comment by lellid** [comment]
I guess, that is the tracking approach from eyalsk? I assume that the intermediates are tracked by the `arena`, too. But now imaging the last line of code is inside a loop with 100 iterations. You can't wait for the entire code to execute and then destroy the arena. And, as I wrote above, it gets much more complicated if a lot of matrix and vector classes are involved.
I does not say that it is impossible to do so, but it would make life easier. And any other good idea how to tackle is problem is equally welcome!

**Comment by CyrusNajmabadi** [comment]
> You can't wait for the entire code to execute and then destroy the arena.

I would make a subarena within that loop.  :)
It would likely help to see an example of what you're talking about and what lifetimes you think are important.

**Comment by lellid** [comment]
Yes, and this is exactly what the compiler should do for me!

**Comment by CyrusNajmabadi** [comment]
> Yes, and this is exactly what the compiler should do for me!

I disagree.  Lifetime management should not be implicit in this fashion.

**Comment by msedi** [comment]
@lellid: We had the same problem and our data is pretty large. We have operations on 3D matrices, where each matrix could have 30GB in size. So keeping the data in memory and also performing operations on the data is pretty heavy.

A simple + operation would mean that another 30GB are created. Its getting worse if you have to perform something like this

`Result = w1 * (A + B) - w2 * (A - B) `

I also strongly disagree with @CyrusNajmabadi. That this is easier to read

`Result = DataMath.Sub(DataMath.Mul(w1, DataMath.Add(A,  B)) - DataMath.Mul(w2, DataMath.Sub((A - B)))`

Which is only a primitive expression, we have even worse, especially when it comes to exponentials, derivatives or interpolations.

Additionally we had the problem that you cannot override operators on interfaces. So we went with the DataMath and we have to stay with this ugly solution until there are no operator overloads on interfaces.

Nevertheless I thought a lot about this and since there is not only the problem of having memory you have to get rid of (collect) after each operation the performance is also a problem, since in our case we have to run many times over the data again and again.

So our initial idea was to resolve the operation at runtime into one closed expression, which is of course true only linear expressions. Nevertheless this is the most common situation.  First we solved this by the old CSharp compiler where we generated code at runtime, with roslyn things are a bit different now but you can do the same things. The draw is that we had to write the math expression as string, which was not nice.

A new idea now is to not return the result directly but to return an expression. The result is evaluated lazily. This means that for each operator we needed to define many variants (see below) including (but not shown) the scalar variants also, which is of course quite some work.

```
DataExpression operator +(DataExpression A, DataExpression B)
DataExpression operator +(Data A, Data B)
DataExpression operator +(DataExpression A, Data B)
DataExpression operator +(Data A, DataExpression B)
Data operator (DataExpression A)
```

With this solution no operation is performed and no data is allocated until you same invoke the operation either by a call or by simply assigning the expression to the target Data instance.

The advantage is also that in linear cases no intermediate data is allocated so for the example with the 30GB and the operation

`Result = w1 * (A + B) - w2 * (A - B) `

we would have and intermediate memory allocation of 120GB which we don't have with the expression. Also the problem of running multiple times from head to toe through the data does not exist anymore. If you assume the overhead of running through the data is (estimated 5s) and in the example above you have to run 5 times over the data, we have an estimated total time of 25s which is reduced to 5s + (some additional overhead).

One problem is though that the inputs may not be disposed until the expression is evaluated.

In my opinion there are still some problems that need support from C# or the runtime (like operators on interfaces and math on generics) but over 10 years of fighting with this problems this was the best we could come up.

Hope this helps.

**Comment by CyrusNajmabadi** [comment]
> I also strongly disagree with @CyrusNajmabadi. That this is easier to read

I never said it was easier to read. :)

**Comment by msedi** [comment]
@CyrusNajmabadi:

> > If you would only have the latter source, could you easily see the expression behind it?
> 
> Yes :)

Seemed like you did ;-) I misinterpreted it in a way that it seemed as you preferred the latter solution ;-)

**Comment by CyrusNajmabadi** [comment]
> Seemed like you did

I didn't :)    I was saying though that i could easily see the expression behind the method call version.  After all, 99%+ of our code is written in method-call form, so seeing structure there really isn't difficult.

I def didn't say it was easier though, :)

I get the desire to continue using basic arithmetic operations.  However, i've now heard two reasons why i think that's a bad idea:

1. lifetime management for this sort of thing should not be implicit.
2. basic operators should not end up being massively expensive (i.e. taking many seconds).

This goes against the desired intent of why operator overloading was provided.  

If you're at the point that you have huge resources with complex ownership and disposal patterns, and you have operations that may end up doing massive amounts of work, my position stands that overloaded operators are an inappropriate solution.  C# very intentionally does not make certain things possible because if they were possible their use would feel like an abuse.  This is one of those cases for me.   :)

**Comment by msedi** [comment]
@CyrusNajmabadi: Yes, I absolutely agree, that this is nothing that should be done implicitly. This is the reason why we switched to expressions and static Math classes that handle this. Here you have more control over the disposal of your data. In our case expressions are a convenient way to solve the problem with math operations, and currently it works quite well, nevertheless with the regular operators there are also several drawbacks:

1. They don't support async/wait
2. No cancellation
3. No progress.

This is way we go currently with the two approaches

1. operators for convenience
2. static math methods for the more fine grained control over what needs to be done.

However I understand the requirement from @lellid to do it. It is more natural to write. The question is if in his case I had the same problem using data from `ArrayPool` or `MemoryPool `but in the end it would help if there would be a deterministic disposal of the class. Or if there could be  stack-type object that has a deterministic finalizer (which is another complex discussion...)

**Comment by svick** [comment]
@msedi 

> Additionally we had the problem that you cannot override operators on interfaces.

You can overload operators on interfaces in C# 8.0. [Here is a dummy example that shows that.](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmARgFgAoagSzwxigDMBDMGAAgEkBZVjFFoJqAb2qdJnIgGZpZJD36DhnCAAcmA6JwDUACj4ChCTgBsYzDGiXHVQgOYALDAEpOAXgB8EqX4tWAHQAggAmofqOLq4A3L6S8bYqpmERRslqGE5MsdQAvtTUpJzKJmQIZJwgSSZiibKc9BicAGqsZjhcopwOMBgxnOpCAG4CXADOfQMFNFR+DaXC5WT6TZyjHTDu3q3tnZ7rezBxs/Ny6SacqYaLphBZOYnic35StMyc+vfZUI3jJXYKhVMj9lq5Eq8pEQAOycPAwADuAOSy30WVo40CbU2ehBTGWWKOuRekIhUiyUAgSPhSO4eA2tFCAHlNFABLQIHgAKIIDjqDAcvD6Yl+GZ+epyCDDJhCUJcIhkAAMnAAKhAAMoqPAOYWeLy7TaBNWaoTa4UnGZFEicAAKdRJ8kURBQAPowqeZMkF1UAFtKh44YjkWUKvoyCLITVfdaAzTg0tQyRiZ7Dr8fXIA37cT6SCcUwqAJz6dMRyQzPJAA==)

> This means that for each operator we needed to define many variants

Wouldn't an implicit conversion operator from `Data` to `DataExpression` help with that?

**Comment by msedi** [comment]
@svick: Thanks for the info. To be honest, we started with C# 4 and we didn't migrate. Nevertheless I knew about the feature but I always forget. I checked with the docs in "What's new in C# 8" there is only a section around default interface methods but nothing that tells that it's also possible with operators.

Sometimes you guys should promote your work more ;-)

**Comment by MillKaDe** [comment]
@msedi 

Huh ? 30 GB per matrix ?  
Are you computing [***The Matrix <sup>(tm)</sup>***](https://en.wikipedia.org/wiki/The_Matrix) itself ?

So you are confronted with the choice between:
- [*blue pill*](https://en.wikipedia.org/wiki/Red_pill_and_blue_pill) : nice readable formulas like `w * (a + b)` with infix operators  
versus
- [*red pill*](https://en.wikipedia.org/wiki/Red_pill_and_blue_pill) : efficient but ugly nested method calls `Mul (w, Add (a, b))`

Why not take both ?  
How ? Use `Expression Trees` from `System.Linq.Expression`.

A *normal function / method* ..
```c#
int AMethod (int a, int b) => a + b;
```
.. can be replaced by a *lambda function* .. 
```c#
Func<int,int,int> AFunc = (a, b) => a + b;
```
.. which can be replaced by an *expression tree* .. 
```c#
Expression<Func<int,int,int>> AExpr = (a, b) => a + b;
```

Note, that the expression `a + b` is **NOT** a string.  So you have full Intellisense support, 
type and syntax checking by the compiler and you can use all your own favourite data types like 
matrix, vector and whatever.

At compile time, the C# compiler parses the expression `a + b` and creates code, that builds 
an `expression tree` at runtime.  That tree looks something like this:
```c#
{root}: Lambda [Func<int,int,int>] // (a, b) => a + b
+ {param1}: Parameter 'a' 
+ {param2}: Parameter 'b' 
+- {body}: Binary.Add // a + b
   + {left}: Parameter 'a' 
   + {right}: Parameter 'b' 
```

At runtime, you can walk and modify that tree - for example you can replace all the beautiful 
*binary operator sub-expressions* with *ugly nested method calls* (see below).
```c#
{root}: Lambda [Func<int,int,int>] // (a, b) => a + b
+ {param1}: Parameter 'a' 
+ {param2}: Parameter 'b' 
+- {body}: Call 'OptAdd' // OptAdd (a, b)
   + {arg1}: Parameter 'a' 
   + {arg2}: Parameter 'b' 
```

When your done with modifying the tree, you compile it to a lambda function and execute it:
```c#
var AFunc2 = AExpr.Compile ();
var result = AFunc2 (2, 3);
```

-----------

Lets start with:
```c#
using System;
using System.Linq;
using System.Linq.Expressions;

namespace Matrix {
  // ...  
} // namespace
```
.. and add way too simple `Matrix` and `vector` types .. 
```c#
public readonly struct V {
  readonly byte X; // size
  readonly int[] v; // X ints
  public V (int i) { X = 1; v = new int[X]; v[0] = i; }
  public int this[int x] => v[x];
  public override string ToString () => $"V[{X}] = {v[0]}";
} // V

public readonly struct M {
  readonly byte X, Y; // sizes in 2 dimensions
  readonly int[,] m; // X * Y ints
  // sizeof (M): 1 + 1 + 8 = 10 bytes plus huuuge array m on heap
  public M (byte x, byte y) { X = x; Y = y; m = new int[x, y]; }
  public M (int i) : this (1, 1) => m[0, 0] = i;
  public int this[int x, int y] => m[x, y];
  public override string ToString () => $"M[{X},{Y}] = {m[0, 0]}";
  public static M operator + (M a, M b) => M.OptAdd (a, b); // forward op to internal function
  public static M operator * (M a, M b) => M.OptMul (a, b);
  public static M operator * (M a, V b) => M.OptMulV (a, b);
  public static M OptAdd (M a, M b) => new M (a[0, 0] + b[0, 0]);
  public static M OptMul (M a, M b) => new M (a[0, 0] * b[0, 0]);
  public static M OptMulV (M a, V b) => new M (a[0, 0] * b[0]);
} // M
```
.. which you later will replace with you huge `Matrix` type.  
I am sure you already avoid storing zero values to bring the size down to 27 GB or so .. ;-P  

Next we add an *expression tree optimizer*
```c#
public static class ExprOptimizer {
  class Optimizer : ExpressionVisitor {
    class Map { // map binary operator to method
      public readonly BinaryExpression BO; // bin op to search for
      public readonly MethodCallExpression MC; // method call to replace bin op
      public Map (BinaryExpression bo, MethodCallExpression mc) { BO = bo; MC = mc; }
      public override string ToString () => $"{BO} --> {MC}";
      public bool Match (BinaryExpression be) // match op and operands
        => BO.NodeType == be.NodeType && BO.Left.Type == be.Left.Type && BO.Right.Type == be.Right.Type;
      public static Map Create<A, B, R> (Expression<Func<A, B, R>> s, Expression<Func<A, B, R>> r)
        => s.Body is BinaryExpression bo && r.Body is MethodCallExpression mc ? new Map (bo, mc) : null;
    } // Map
    static readonly Map[] mappings = new Map[] { // map op to func
      Map.Create ((M x, M y) => x + y, (M x, M y) => M.OptAdd (x, y)),
      Map.Create ((M x, M y) => x * y, (M x, M y) => M.OptMul (x, y)),
      Map.Create ((M x, V y) => x * y, (M x, V y) => M.OptMulV (x, y)),
    };
    protected override Expression VisitBinary (BinaryExpression be) {
      var m = mappings.FirstOrDefault (m => m.Match (be));
      return m is null ? base.VisitBinary (be) :
        Expression.Call (null, m.MC.Method, Visit (be.Left), Visit (be.Right));
    }
  } // Optimizer
  public static Expression<F> Optimize<F> (this Expression<F> el) 
    => (new Optimizer ().Visit (el)) as Expression<F>;
} // ExprOptimizer
```
The optimizer is called with an *expression tree* representing a *lambda function* with 
two (or three) parameters.
`Optimize` optimizes the lambda function body by recursively replacing 
all `a + b` expressions with method calls `OptAdd (a, b)`, 
and `a * b` with `OptMul (a, b)`.

Finally create a helper class for wrapper functions (which hides all the nasty details) :
```c#
public class EFunc {
  public class FuncN<F> { // base class for expression-wrapper funcs
    protected readonly F f; // optimized expression compiled to function
    public FuncN (Expression<F> e) => f = e.Optimize ().Compile ();
  }
  public class Func3<A, B, C, R> : FuncN<Func<A, B, C, R>> { // 3 args
    public Func3 (Expression<Func<A, B, C, R>> e) : base (e) { }
    public R Call (A a, B b, C c) => f (a, b, c);
  }
  public class Func2<A, B, R> : FuncN<Func<A, B, R>> { // 2 args
    public Func2 (Expression<Func<A, B, R>> e) : base (e) { }
    public R Call (A a, B b) => f (a, b);
  }

  public static Func2<A, B, R> Expr<A, B, R> (Expression<Func<A, B, R>> e) 
    => new Func2<A, B, R> (e);
  public static Func3<A, B, C, R> Expr<A, B, C, R> (Expression<Func<A, B, C, R>> e) 
    => new Func3<A, B, C, R> (e);

  public static R Call<A, B, R> (A a, B b, Expression<Func<A, B, R>> e) 
    => new Func2<A, B, R> (e).Call (a, b);
  public static R Call<A, B, C, R> (A a, B b, C c, Expression<Func<A, B, C, R>> e) 
    => new Func3<A, B, C, R> (e).Call (a, b, c);
}
```
and use it :
```c#
public class MatrixTest {
  public void Test () {
    M a = new M (2); // get/read data
    M b = new M (3);
    M w = new M (4);
    V v = new V (5);

    // 1) create wrapper funcs and store as delegate
    var Add = EFunc.Expr ((M x, M y) => x + y);
    var Mul = EFunc.Expr ((M x, M y) => x * y);
    var MulV = EFunc.Expr ((M x, V y) => x * y);
    var AddMul = EFunc.Expr ((M x, M y, M z) => z * (x + y));

    // 2) call wrapper func via delegate
    var xAdd_5 = Add.Call (a, b); // 2 + 3 = 5
    var xMul_6 = Mul.Call (a, b); // 2 * 3 = 6
    var xMulV_10 = MulV.Call (a, v); // 2 * 5 = 10
    var xAddMul_20 = AddMul.Call (a, b, w); // 4 * (2 + 3) = 20

    // 3) create wrapper func and directly call it (without storing in delegate)
    var yAdd_5 = EFunc.Expr ((M x, M y) => x + y).Call (a, b); // 2 + 3 = 5
    var yMul_6 = EFunc.Expr ((M x, M y) => x * y).Call (a, b); // 2 * 3 = 6
    var yMulV_10 = EFunc.Expr ((M x, V y) => x * y).Call (a, v); // 2 * 5 = 10
    var yAddMul_20 = EFunc.Expr ((M x, M y, M z) => z * (x + y)).Call (a, b, w); // 4 * (2 + 3) = 20

    // 4) if Expr() and Call() parameter lists are merged, the compiler can infer parameter types M and V
    var zAdd_5 = EFunc.Call (a, b, (x, y) => x + y); // 2 + 3 = 5
    var zMul_6 = EFunc.Call (a, b, (x, y) => x * y); // 2 * 3 = 6
    var zMulV_10 = EFunc.Call (a, v, (x, y) => x * y); // 2 * 5 = 10
    var zAddMul_20 = EFunc.Call (a, b, w, (x, y, z) => z * (x + y)); // 4 * (2 + 3) = 20
  }
} // class MatrixTest
```

Note, that optimizing and compiling the expression tree happens just **ONCE** per formula, if you store it in a delegate.

-----------
In your application, the `OptAdd` method in type `M`atrix can do local optimizations 
per single operation - which are not possible with the operator.  Use this for optimizations at the innermost level.

For more complex optimizations, put the modified *expression tree* compiled-to-func in a 
wrapper function, so you can do cleanup stuff before and/or after.  Use this for optimizations at the outermost level.

If you need optimizations somewhere between inner- and outermost, you can modify the expression tree.

----------------------
Updates:
- bug fix in ExprOptimizer
- EFunc wrappers added

**Comment by msedi** [comment]
@MillKaDe: Wow, thanks for the detailed description. This is exactly what I was thinking of.

BTW: Nowadays our matrices are much bigger. We are getting into the range of 100GB...

**Comment by MillKaDe** [comment]
Some hints for **ET** (Expression Trees):

NET 3.5 has **ETv1** (just expressions), NET 4.0 has **ETv2** which adds statements and many 
other stuff - see spec below for details.

The guru / mastermind behind ET seems to be *Bart De Smet*.  
His [GitHub](https://github.com/bartdesmet) has later ET versions convering the newer C# stuff.

Some useful links:
- Spec: [Expression Tree Specification](https://github.com/IronLanguages/dlr/tree/master/Docs)
- Blog: [Bart De Smet's blog](http://community.bartdesmet.net/blogs/bart/default.aspx)
- Book: [C# Unleashed 5, Bart De Smet](https://www.informit.com/store/c-sharp-5.0-unleashed-9780133391497)

100 GB matrices - what the hell are you computing ??  
Anyway, you have two problems: space and time .. ;-P

Some wild ideas:
- Use smaller precision type ? (32 bit float or even 16 bit)
- Avoid storing zero values ?
- Shortcut calculations with zero ? (because `a + 0 = a` and `a * 0 = 0`)
- Treat 'very small' values as zero ?
- Use fast CPU support: [Intrinsics.X86](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.intrinsics.x86?view=netcore-3.1)
- Parallel computation ?
- Use GPU for number crunching ? [CUDA](https://en.wikipedia.org/wiki/CUDA), [OpenCL](https://en.wikipedia.org/wiki/OpenCL)
- Speed up innermost loops / fundamental operations ? (unsafe C#, native C, hand-tuned assembler)

**Comment by msedi** [comment]
@MillKaDe: Thanks for the infos, they are very welcome.
Regarding your suggestions:

1. It is noise X-ray data where the precision is necessary (float is sufficient though)
2. Space is not such a big problem our machine has 1TB RAM
3. Additionally where are using our own GC like methods (memory mapped files and a few things more)
4. We have not so many zero values since there is poisson noise and electronic noise added to the data, so we have almost no 0 values in the data.
5. "Treat small values as zero", as said above we need the precision but we already have cut-off values that do something similar.
6. We already do use intrinsics, sometimes using Vector<T> sometimes using the more "hardcore" intrinsics
7. Parallel: We already do
8. We are also using CUDA with textures, layered textures, etc. OpenCL turned oiut to be a problem on systems (GPUs) that do not support textures (there is no fallback and it simoply crashes). That's why we are currently bound to CUDA only which I don't like because we need to use a CUDA-based GPU. But it's currently the best and only solution. Sometimes we are also using MKL.
9. Speed-up: already done with unsafe but without assembler since then we are so platform dependent and in the past C# never caused a problem, but we have many problems with  C++ CLI projects that are 1) slowing down the build process tremendously and 2.) are always causing troubles with the C# interop. e.g. currently with the new version of VS there's a problem with readonly structs that do not allow interfaces anymore.

Nevertheless: Thanks for sharing your thoughts.
